### ISSUE TITLE
Placeholder scope/visibility

### ISSUE THREAD

**Issue by dy** [mindeavor/es-pipeline-operator#262]
Following #261 − what are the boundaries for placeholder? (sorry if question was asked.)

```js
// is it possible to nest placeholder within a function?
source |> a => # + a |> output(#)

// is it possible to define a function outside of pipeline to allow reusing?
const add= a => # + a
source |> add(a) |> output(#)

// is placeholder available from async context?
source |> setTimeout(() => # + 1) 

// will a function defined in pipeline context retain reference to placeholder?
let add = source |> a => # + a
add(1)
```

**Comment by ljharb** [comment]
Yes, no, yes, yes.

**Comment by dy** [comment]
```
// is placeholder scope defined by brackets?
source |> a + # |> b + #
source |> (a + # |> b) + #
source |> a + (# |> b + #)
```

**Comment by ljharb** [comment]
Each pipeline step needs a placeholder, and i would assume that a pipeline, being an expression, would be entirely contained  by parens - meaning, your third example there seems like it might not parse.

**Comment by dy** [comment]
Potential userland scenario: since functions are not reusable, fake stack can be organized in underscore chain fashion:

```js
let $
const _ = arg => $ = arg,
    add = a => $ = $ + a,
    output = () => out($)
    
_(source) |> add(a) |> output()
```

**Comment by dy** [comment]
> would be entirely contained by parens - meaning, your third example there seems like it might not parse.

Does it mean parens are limitations to placeholder visibility? In that case boundaries should be more clearly defined...

```js
// if that's not valid
source |> a + (# |> b + #) 

// then what about
source |> a + (a => # |> b + a)(#)
source |> a + (a => (# |> b + a))(#)
value |> foo(() => class { constructor() { return addEvent('x', function () { return () => [await x`${ # }`]})} } )
```

**Comment by ljharb** [comment]
No, it’s saying that the inner pipeline could only parse in its entirety because it’s inside parens. Parens don’t affect placeholder visibility.

**Comment by ljharb** [comment]
although, it also might parse so the LHS of the inner pipeline is the outer pipeline’s placeholder?

**Comment by js-choi** [comment]
Keep in mind the following paragraphs from the [explainer’s description](https://github.com/tc39/proposal-pipeline-operator/#description):

> A pipe body must use its topic value at least once. For example, `value |> foo + 1` is invalid syntax, because its body does not contain a topic reference. This design is because omission of the topic reference from a pipe expression’s body is almost certainly an accidental programmer error.
>
> Likewise, a topic reference must be contained in a pipe body. Using a topic reference outside of a pipe body is also invalid syntax.
>
> To prevent confusing grouping, it is invalid syntax to use other operators that have the same precedence (the arrow `=>`, the ternary conditional operator `? :`, the assignment operators, and the yield operator) as a pipe head or body. When using `|>` with these operators, we must use parentheses to explicitly indicate what grouping is correct. For example, `a |> b ? % : c |> %.d` is invalid syntax; it should be corrected to either `a |> (b ? % : c) |> %.d` or `a |> (b ? % : c |> %.d)`.

This is formally defined in [this early-errors section of the specification](https://tc39.es/proposal-pipeline-operator/#sec-pipe-operator-static-semantics-early-errors).

So the lines in each of the following code blocks are equivalent to one another.

```js
source |> a => # + a |> output(#)
SyntaxError: Unparenthesized arrow function found in pipe body. Arrow functions must be parenthesized inside pipe body.
```

```js
const add= a => # + a
SyntaxError: Unbound topic # was found. Topics may only be used within a pipe body.
```

```js
source |> add(a) |> output(#)
SyntaxError: Topic was not used within pipe body `add(a)`.
```

```js
source |> setTimeout(() => # + 1) 
setTimeout(() => source + 1)
```

```js
_(source) |> add(a) |> output()
SyntaxError: Topic was not used within pipe body `add(a)`.
```

```js
let add = source |> a => # + a;
SyntaxError: Unparenthesized arrow function found in pipe body. Arrow functions must be parenthesized inside pipe body.
```

```js
source |> a + # |> b + #
source |> (a + #) |> (b + #)
b + (a + source)
```

```js
source |> (a + # |> b) + #
SyntaxError: Topic was not used within pipe body `b`.
```

```js
source |> a + (# |> b + #)
source |> a + (b + #)
a + (b + source)
```

```js
source |> a + (# |> b + #)
source |> (a + #) |> (b + #)
b + (a + source)
```

```js
source |> a + (a => #.x |> b + a)(#)
SyntaxError: Topic was not used within pipe body `b + a`.
```

```js
source |> a + (a => #.x |> b + a + #)(#)
source |> a + (a => b + a + (#.x))(#)
a + (a => b + a + source.x)(source)
```

```js
source |> a + (a => (# |> b + a))(#)
SyntaxError: Topic was not used within pipe body `b + a`.
```

```js
value |> foo(() => class { constructor() { return addEvent('x', function () { return () => [await x`${ # }`]})} } )
foo(() => class { constructor() { return addEvent('x', function () { return () => [await x`${ value }`]})} } )
```

In general, I would encourage developers to keep their pipe expressions’ bodies as small, simple, and readable as possible. In particular, I would avoid nesting parenthesized pipe expressions: the purpose of pipes is to untangle nested expressions, not to create more.

**Comment by js-choi** [comment]
> Ternary has [higher precedence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence) than assignment, I guess there should not be a problem with `a |> b ? % : c`.

I believe this is an error with MDN’s article. The [actual language specification defines them on the same precedence level](https://tc39.es/ecma262/#prod-AssignmentExpression). I will try to open a pull request to this article later.

**Comment by dy** [comment]
They seem to indeed have rather same precedence:
```
let a, b, c
c = 1 ? a : b = 3
c // undefined
// if ?: had higher precedence than assignment, it'd evaluate as
// c = (1 ? a : b) = 3 →  c = (a) = 3  /
// but it evaluates as
// c = 1 ? a : (b = 3)
```

**Comment by js-choi** [comment]
This is a little off topic, but the weirdness here is that `=` is not really a binary operator in the usual sense. It is a weird operator: it is kind of like a prefix operator that operates on its RHS. Its LHS is not an ordinary expression but rather more like a parameter list.

**Comment by dy** [comment]
But partially precedence is applicable to `=` operator with combination with others:
```
a += 1
a.b += 1
((a.b)) += 1
a?.b += 1 // breaks
```
Sort-of LHS in `=` should not include anything calculable.

**Comment by dy** [comment]
Speaking of precedence, to clarify:
```
a ? b : c |> #
// is equivalent to
(a ? b : c) |> #
// and not
a ? b : (c |> #)
```
?

**Comment by js-choi** [comment]
I think we intend to ban `a ? b : c |> #` without parentheses and to force the developer to specify between `(a ? b : c) |> #` or `a ? b : (c |> #)`. I forgot to put this particular error in the spec, though, and I should add it. We should also clarify this in the explainer.
